### PR TITLE
New Matcher API

### PR DESCRIPTION
**Pull request by jeffh**
Open for feedback, although there still is more work required. The high level goals are:

 - Redesign the Matcher API to be easier to use and help with common pitfalls
    - Nil checking
    - Failure message reporting for `to` and `toNot` implementations
    - Maintain correct behavior if given to another matcher
 - Allow matchers that use other matchers to work more seamlessly
    - The differing types the matchers currently return make it difficult to write composable matchers.

Implementation-wise, we're replacing all matcher types:

 - `Matcher`
 - `MatcherFunc`
 - `NonNilMatcherFunc`

...into the same type (currently called `Predicate`). This is a breaking change, so it is probably best if we can change API in the following ways:

 - Help handle nil-failure checking automatically
 - Help ensure that failure messages are always generated by the matcher
    - It's needed for both `to` and `toNot` expectations 
    - Higher-order matchers will typically use the error messages from lower-order matchers (e.g. - `allPass`)
 - Simplify `FailureMessage` to be more explicit and cover the common cases:
    - Fail based on invalid usage of the matcher (e.g. - `equal(nil)`)
    - Fail while displaying the actual & expected values (e.g. - `equal`)
    - Fail while displaying only the expected value / result (e.g. - `throw`)
    - Fail while displaying a suggestion why it failed (e.g. - hitting at using `beNil()` if actual value is nil)
    - Fail while displaying more comprehensive information
    - Fail while displaying user-provided information

The goal is that consumers can migrate over to the new matcher APIs in steps (which will probably be two major versions).

 - The first release will introduce the new matcher api and an easy compatibility layer to "upgrade" quickly (using the `Matcher.predicate`). The old API will be marked as deprecated (aka - print warnings)
 - The second release will remove the old API.

TODO:

 - [x] Fix failing tests
    - [x] `FailureMessage` to `PredicateResult` conversion is not implemented
 - [x] Change `Predicate` constructors to not need explicit return values
 - [ ] ~~Rename `Predicate` to a better name?~~
 - [x] Remove any `.predicate` conversions (but keep that method available for migration)
 - [x] Try and revert change: `#define NIMBLE_EXPORT FOUNDATION_EXPORT NIMBLE_OVERLOADABLE`
 - [x] Make sure there's a migration path for `AsyncMatcherWrapper` 
 - [x] Documentation
    - [x] New Matcher API
    - [x] Migrating old matchers

-------------------

Checklist - While not every PR needs it, new features should consider this list:

 - [x] Does this have tests? Keeping same test suite
 - [x] Does this have documentation? Yes
 - [x] Does this break the public API (Requires major version bump)? **Yes**
 - ~~Is this a new feature (Requires minor version bump)?~~

### CODE REVIEW

**Comment by shaps80** [comment]
@jeffh 
When are you planning to get this in? I wanted to work on #218 -- perhaps I should merge this change in to mine to include it?

**Comment by jeffh** [comment]
Yeah, I'm planning on working on it this Sunday and Monday. I'll probably
have to cut its scope to get it in a workable state.

**Comment by shaps80** [comment]
Fair enough. As you said this is definitely a change worth getting in sooner rather than later, and then I can tackle the PO Matchers ticket, for both Nimble and Spry.

**Comment by jeffh** [comment]
@Quick/core, this should be mostly ready for review! I still need to write documentation and to plan a migration strategy for `AsyncMatcherWrapper` (probably remove it as a `Matcher` / `Predicate`).

File: Sources/Nimble/Expectation.swift [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import Foundation
 
+// Deprecated
```
Comment (shaps80) [inline Sources/Nimble/Expectation.swift]: Any reason why you don't want use Swift attributes?

File: Sources/Nimble/Matchers/Predicate.swift [hunk 2]
```diff
@@ -9,107 +9,130 @@
...
+        return Predicate<T> { actual in
+            return PredicateResult(status: try matcher(actual), message: .ExpectedActualValueTo(msg))
         }
     }
 }
 
-public enum ExpectationStyle {
+// Question: Should this be exposed? It's safer to not for now and decide later.
+internal enum ExpectationStyle {
     case ToMatch, ToNotMatch
```
Comment (shaps80) [inline Sources/Nimble/Matchers/Predicate.swift]: I haven't pulled this branch -- is this not Swift 3?

File: Sources/Nimble/Expectation.swift (hunk 1 above)
Comment (jeffh) [inline Sources/Nimble/Expectation.swift]: I'd have to check if these attributes still cause build errors. If `@available(*, deprecated, message: "...")` causes a build failure, then it's effectively the same as removing the API.

File: Sources/Nimble/Matchers/Predicate.swift (hunk 2 above)
Comment (jeffh) [inline Sources/Nimble/Matchers/Predicate.swift]: Agreed. I'll get that change in

File: Sources/Nimble/Expectation.swift (hunk 1 above)
Comment (shaps80) [inline Sources/Nimble/Expectation.swift]: `deprecated` causes a warning on calling code, not a failure

**Pull request merged**